### DISCUSSION TITLE
GoDroid versus LineageOS... on Pi 400 !

### DISCUSSION THREAD

**Discussion by DORNINEM** [tesla-android/issue-tracker#46] [Q&A] [answered]
I'm wondering why Tesla-Android is using GoDroid instead of LineageOS9 (same Android 12L)
LineAge is much more easy to install and is supporting many Raspberry family hardware, especially the Raspberry Pi400.
GoDroid only supports Pi4, no network interface (RJ45 or wi-fi) on my Pi400 with GoDroid :( 

Pi400 is wonderful because quite same hardware but not same hardware : CPU frequency is 1800 mhz and can be overclocked to 2000. No need to use a fan or something like this thanks  to the board design and the use of aluminium heat dispatcher .
Yes the Pi400 is inside a keyboard... but it fits like a charm under Tesla's a seat .
Some links about Pi400 : 
https://www.xda-developers.com/raspberry-pi-4-b-pi-400-compute-model-4-android-12-port/
https://konstakang.com/devices/rpi4/LineageOS19/
https://all3dp.com/2/raspberry-pi-400-vs-raspberry-pi-4-differences/
https://tutorial.cytron.io/2020/11/02/raspberry-pi-400-thermal-performance/

Regards.
Emmanuel (from France !)

**Comment by mikegapinski** [comment, marked as answer]
The owner of Lineage-RPI does not share source code for device configs.

Besides that most of the stuff in Lineage comes from Glodroid anyways.

The install process will be simplified, don't worry 😃

**Reply by DORNINEM** [reply]
Mike,
No problems regarding the setup process.. it's not so easy as LineageOS but for a alpha or beta release this is not an issue.
I tried to gather some information from Konsta (the guy who maintain the LineageOS Raspberry project) and ... "Complete device configuration for my more recent builds is not available" and  "My Raspberry Pi device specific configuration is my own effort (I forked from android-rpi several years/Android version ago and I've maintained it independently since)" :(
As the hardware of a PI400 is very similar to a PI4 I think that it should be easy to fix for a AOSP geek. Wi-Fi or RJ45 Ethernet port are not functionnal... but the bluetooth interface is working. 
I do some benchmarks  and I really think that the PI400 is a really nice hardware solution for TeslaAndroid, faster and no thermal issue :)    

Regarding the future of your project I don't really understand the need of the LTE hardware? Why don't you use the iPhone connection sharing?

**Reply by mikegapinski** [reply]
The next build should run on the 400, I've added the patches to my device tree

**Reply by DORNINEM** [reply]
Fine, thanks ;) 
I will test it as soon as you will release the version ! With feedbacks of course.

**Reply by focher** [reply]
I was curious about the same regarding the use of separate LTE hardware as opposed to the phone's personal hotspot functionality. I imagine that would only work for a wired connection, because at least on an iPhone the wifi cannot be used both for a personal hotspot and to connect to a different hotspot (ie Tesla Android).

**Reply by mikegapinski** [reply]
@focher Using an iPhone as an external internet source was very unreliable in my testing. That won't be an issue however since I've managed to bypass the requirement of having an internet connection in order for the car to maintain Wi-Fi connection. When the next version goes live an LTE dongle will be an optional add-on that unlocks more functionality

**Reply by GtHbbrr** [reply]
I just ordered a PI400. What is the current situation for installing TeslaAndroid on thePI400?

**Reply by DORNINEM** [reply]
Works fine 
I think better than a regular Pi4, I overclocked to 2ghz and no lag, no heat

**Reply by GtHbbrr** [reply]
How do you connect your PI400 to internet?

**Comment by DORNINEM** [comment]
@mikegapinski :  the version 2022.25.1 works like a charm on my Pi400 !  Thanks a lot.
